src/components/Table/Table.js

```diff
@@ -8,99 +8,216 @@ import TableRow from '@mui/material/TableRow';
+        fetch(`http://localhost:${process.env.REACT_APP_SERVER_PORT}`)
+            .then((response) => response.json())
+            .then((data) => {
+                const currenciesList = Object.entries(data).map(([name, value]) => ({
+                    currency: name,
+                    value,
+                    isSelected: false,
+                }))
+                setCurrencies(currenciesList)
+
+            })
+    }
+
+    useEffect(() => {
+
+        getCurrenciesData()
+
+        const getSubscribedCurrencies = async () => {
+            const existingSubs = await getSubscriptionUrl()
+            await getSubscribedCurrenciesAPIRequest(existingSubs.endpoint)
+        }
+
+        if (!("Notification" in window)) {
+            alert("This browser does not support desktop notification");
+        } else if (Notification.permission === pushNotificationPermitStatuses.granted) {
+            getSubscribedCurrencies()
+        }
+
+        broadcastChannel.onmessage = (event) => {
+            const {type, payload} = event.data;
+
+            if (type === broadcastChannelTypes.getPushMessage) {
+
+                const [currency, value] = payload.split(':')
+                setUpdatedCurrencies([currency])
+
+                setCurrencies(prev => prev.map(item => {
+                    if (item.currency === currency) {
+                        return {
+                            ...item, value: value.trim()
+                        }
+                    }
+                    return item
+                }))
+            }
+        }
+
         setLoading(false);
-      });
-  }, []);
-
-  const handleSubscribeToCurrencyUpdate = (currencyName) => {
-    setCurrencies(
-      currencies.map((item) => (
-        currencyName === item.currency ? { ...item, isSelected: !item.isSelected } : item
-      )),
+
+    }, []);
+
+
+    const updateSubscriptionAPIRequest = async (currencies) => {
+
+        const subscription = await getSubscriptionUrl()
+
+        fetch(`http://localhost:${process.env.REACT_APP_SERVER_PORT}/subscription`, {
+            method: "post",
+            headers: {"Content-type": "application/json"},
+            body: JSON.stringify({subscription, currencies})
+        }).catch(error => {
+            console.log("Error", error);
+        })
+    }
+
+    const handleCheckbox = (currencyName) => {
+        const updatedCurrencies = currencies.map((item) => (
+            currencyName === item.currency ? {...item, isSelected: !item.isSelected} : item
+        ))
+
+        setCurrencies(updatedCurrencies);
+
+        const selected = updatedCurrencies.filter(item => item.isSelected).map(cur => cur.currency)
+        updateSubscriptionAPIRequest(selected)
+    }
+
+    const handleFirstGrant = (currencyName) => {
+        handleCheckbox(currencyName)
+
+        broadcastChannel.postMessage({type: broadcastChannelTypes.permissionGranted})
+
+        broadcastChannel.onmessage = event => {
```

Comment:
I think we can move handling of all messages from channel in one use effect and use `switch/case`